server/src/main/java/org/elasticsearch/search/SearchHit.java

```diff
@@ -653,13 +704,72 @@ public Map<String, Float> getMatchedQueriesAndScores() {
      * @return Inner hits or <code>null</code> if there are none
      */
     public Map<String, SearchHits> getInnerHits() {
+        assert hasReferences();
```

Comment:
Maybe not right? If we do that here, then we'd have to do it for the source itself before returning it and spots like that also? Seems like consistently asserting ref-counts on fields of ref-counted stuff is going to be a lot of code and doesn't add that much since whoever gets this map is probably going to do something with its contents anyway?